src/node_file.cc

```diff
@@ -1150,28 +1153,133 @@ static void RMDir(const FunctionCallbackInfo<Value>& args) {
+    switch (err) {
+      case 0:
+        break;
+      case UV_ENOENT: {
+        std::string dirname = next_path.substr(0, path.find_last_of("\\/"));
+        continuation_data.PushPath(next_path);
+        continuation_data.PushPath(dirname);
+        break;
+      }
+      case UV_EPERM: {
+        return err;
+        break;
+      }default:
+        err = uv_fs_lstat(loop, req, next_path.c_str(), nullptr);
+        if (err < 0) return err;
+        if (!S_ISDIR(req->statbuf.st_mode)) return UV_EEXIST;
+        break;
+    }
+  }
+
+  return 0;
+}
+
+int MKDirpAsync(uv_loop_t* loop,
+                uv_fs_t* req,
+                const char* path,
+                int mode,
+                uv_fs_cb cb) {
+  FSReqBase* req_wrap = static_cast<FSReqBase*>(req->data);
+  // on the first iteration of algorithm, stash state information.
+  if (req_wrap->continuation_data == nullptr) {
+    req_wrap->continuation_data = std::unique_ptr<FSContinuationData>{
+                                         new FSContinuationData(req, mode, cb)};
+    req_wrap->continuation_data->PushPath(path);
+  }
+
+  // on each iteration of algorithm, mkdir directory on top of stack.
+  std::string next_path = req_wrap->continuation_data->PopPath();
+  ssize_t err = uv_fs_mkdir(loop, req, next_path.c_str(), mode,
+                        uv_fs_callback_t{[](uv_fs_t* req) {
+    FSReqBase* req_wrap = static_cast<FSReqBase*>(req->data);
+    Environment* env = req_wrap->env();
+    uv_loop_t* loop = env->event_loop();
+
+    switch (req->result) {
+      case 0: {
+        if (req_wrap->continuation_data->paths.size() == 0) {
+          req_wrap->continuation_data->Done(0);
+        } else {
+          MKDirpAsync(loop, req, req->path,
+                      req_wrap->continuation_data->mode, nullptr);
+        }
+        break;
+      }
+      case UV_ENOENT: {
+        std::string path = req->path;
+        std::string dirname = path.substr(0, path.find_last_of("\\/"));
+        req_wrap->continuation_data->PushPath(path);
+        req_wrap->continuation_data->PushPath(dirname);
+        MKDirpAsync(loop, req, req->path,
+                    req_wrap->continuation_data->mode, nullptr);
+        break;
+      }
+      case UV_EPERM: {
+        req_wrap->continuation_data->Done(req->result);
+        break;
+      } default:
+        if (req->result == UV_EEXIST &&
+            req_wrap->continuation_data->paths.size() > 0) {
+          MKDirpAsync(loop, req, req->path,
+                      req_wrap->continuation_data->mode, nullptr);
+        } else {
+          // verify that the path pointed to is actually a directory.
+          int err = uv_fs_lstat(loop, req, req->path,
+                                uv_fs_callback_t{[](uv_fs_t* req) {
+            FSReqBase* req_wrap = static_cast<FSReqBase*>(req->data);
+            ssize_t err = req->result;
+            if (err == 0 && !S_ISDIR(req->statbuf.st_mode)) err = UV_EEXIST;
```

Comment:
extreme nit: We *generally* put the body of if statements on a new line or within brackets.